controllers/alert_controller.go

```diff
@@ -99,95 +100,80 @@ func (r *AlertReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts Aler
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=alerts/status,verbs=get;update;patch
 
 func (r *AlertReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
-	start := time.Now()
+	reconcileStart := time.Now()
 	log := ctrl.LoggerFrom(ctx)
 
-	alert := &v1beta1.Alert{}
-	if err := r.Get(ctx, req.NamespacedName, alert); err != nil {
+	obj := &apiv1.Alert{}
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// record suspension metrics
-	r.RecordSuspend(ctx, alert, alert.Spec.Suspend)
-
-	if alert.Spec.Suspend {
-		log.Info("Reconciliation is suspended for this object")
-		return ctrl.Result{}, nil
-	}
-
-	patchHelper, err := patch.NewHelper(alert, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	// Initialize the runtime patcher with the current version of the object.
+	patcher := patch.NewSerialPatcher(obj, r.Client)
 
 	defer func() {
-		patchOpts := []patch.Option{
-			patch.WithOwnedConditions{
-				Conditions: []string{
-					meta.ReadyCondition,
-					meta.ReconcilingCondition,
-					meta.StalledCondition,
-				},
-			},
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
+
+		// Emit warning event if the reconciliation failed.
+		if retErr != nil {
+			r.Event(obj, corev1.EventTypeWarning, meta.FailedReason, retErr.Error())
 		}
 
-		if retErr == nil && (result.IsZero() || !result.Requeue) {
-			conditions.Delete(alert, meta.ReconcilingCondition)
-
-			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
-
-			readyCondition := conditions.Get(alert, meta.ReadyCondition)
-			switch readyCondition.Status {
-			case metav1.ConditionFalse:
-				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
-				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
-			case metav1.ConditionTrue:
-				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
-				conditions.Delete(alert, meta.StalledCondition)
-			}
+		// Log and emit success event.
+		if retErr == nil && conditions.IsReady(obj) {
```

Comment:
This check fails even if the main reconciliation operations were successful and Ready=True because `IsReady()` checks if there's any Stalled or Reconciling condition. And the stale Reconciling condition, which is set at the beginning of `reconcile()` is only removed in the `patch()` that's called below at the very end.

Reply:
Yeap, I'll move patch first, like in KC, this way the error event will capture the agregate.